rfcs/0119-testing-conventions.md

```diff
@@ -0,0 +1,105 @@
+---
+feature: Defined conventions around testing of official Nixpkgs packages.
+start-date: 2021-12-29
+author: Jonathan Ringer
+co-authors:
+shepherd-team: @Mic92 @Artturin @kevincox
+shepherd-leader: @Mic92
+related-issues:
+  - [RFC 0088 - Nixpkgs Breaking Change Policy](https://github.com/NixOS/rfcs/pull/88)
+---
+
+# Summary
+[summary]: #summary
+
+When updating or modifying packages, several conventions for testing regressions
+have been adopted. However, these practices are not standard, and generally it's not well
+defined how each testing method should be implemented. It would be beneficial to have
+an unambiguous way to say that a given package, and all downstream dependencies, have
+had as many automated test ran possible. This will give a high degree of certainty that
+a given change is less likely to manifest regressions once introduced on a release
+channel.
+
+Another desire of this rfc is also to have a way for various review tools
+(e.g. ofborg, hydra, nixpkgs-review) to have a standard way to determine if a
+package has additional tests which can help verify its correctness.
+
+# Motivation
+[motivation]: #motivation
+
+Breakages are a constant painpoint for nixpkgs. It is a very poor user experience to
+have a configuration broken because one or more packages fail to build. Often when
+these breakages occur, it is because the change had a large impact on the entirety
+of nixpkgs; and unless there's a dedicated hydra jobset for the pull request, it's
+infeasible to expect pull request authors to verify every package affected
+by a change they are proposing. However, it is feasible to specify packages that
+are very likely to be affected by changes in another package, and use this information
+to help mitigate regressions from appearing in release channels.
+
+# Detailed design
+[design]: #detailed-design
+
+Standardize `passthru.tests.<name>` as a mechanism of 
```

Comment:
The `<name>` syntax suggests that the namespace is flat, but it isn't. Both ofborg and `nix-build` respect the `recurseForDerivations` attribute (aka `recurseIntoAttrs`), which is a valuable behavior
 - for authors, adding an attribute is easier than merging attribute sets
 - for "testers", the computation of the set of attribute names is lazier, improving evaluation performance when they only run a specific test or set of tests

`passthru` is an implementation detail of `mkDerivation` that must not be part of this specification. You could mention it _once_ as a note to guide package authors, and you can use it in examples with `mkDerivation`, but it must not be part of the package specification. (See also for an attempt to define "package" https://github.com/NixOS/nix/issues/6507)

```suggestion
Standardize `<pkg>.tests` as a mechanism of 
```

Reply:
won't `$tests` be populated in each build environment if this is true? the inclusion of `passthru` was because those get pruned, and is the current convention. (not to say that current conventions aren't flawed, and should be changed)